opal/class/opal_lifo.h

```diff
@@ -104,8 +104,47 @@ static inline bool opal_lifo_is_empty( opal_lifo_t* lifo )
     return (opal_list_item_t *) lifo->opal_lifo_head.data.item == &lifo->opal_lifo_ghost;
 }
 
+#if OPAL_HAVE_ATOMIC_LLSC_PTR
 
-#if OPAL_HAVE_ATOMIC_CMPSET_128
+static inline opal_list_item_t *opal_lifo_push_atomic (opal_lifo_t *lifo,
+                                                       opal_list_item_t *item)
+{
+    opal_list_item_t *next;
+
+    do {
+        item->opal_list_next = next = (opal_list_item_t *) opal_atomic_ll_ptr (&lifo->opal_lifo_head.data.item);
+        opal_atomic_wmb ();
+    } while (!opal_atomic_sc_ptr(&lifo->opal_lifo_head.data.item, item));
+
+    opal_atomic_wmb();
+
+    return next;
+}
+
+/* Retrieve one element from the LIFO. If we reach the ghost element then the LIFO
+ * is empty so we return NULL.
+ */
+static inline opal_list_item_t *opal_lifo_pop_atomic (opal_lifo_t* lifo)
+{
+    opal_list_item_t *item, *next;
+
+    do {
+        item = (opal_list_item_t *) opal_atomic_ll_ptr (&lifo->opal_lifo_head.data.item);
+        if (&lifo->opal_lifo_ghost == item) {
+            return NULL;
+        }
+
+        next = (opal_list_item_t *) item->opal_list_next;
+        opal_atomic_isync ();
```

Comment:
Yup. I don't think a barrier is needed there. Was supposed to be removed before I pushed this for PR. I will update the PR to remove this barrier.

As for the isync abstraction. Mellanox pushed the abstraction a little while ago.